packages/hdwallet-provider/src/index.ts

```diff
@@ -47,6 +48,7 @@ class HDWalletProvider {
       shareNonce = true,
       derivationPath = `m/44'/60'/0'/0/`,
       pollingInterval = 4000,
+      chainId,
```

Comment:
So this requires the `chainId` to be passed to the constructor?  Will that require any additional work on the part of the user or do we have things set up so that should happen automatically?

Reply:
They will need to enter it into the config. If they don't, they will just sign transactions in an EIP-155 non-compliant manner. This won't matter when they are using something like Ganache with unlocked accounts because it doesn't check the signature.

Reply:
Oh, that's kind of annoying.  Shouldn't HDWalletProvider be able to determine it itself, rather than requiring some new extra step for people to do?  I don't know, maybe we're OK with this... @gnidan, you have an opinion?

Reply:
I guess it is plausible that we could include an http lib in hdwallet-provider and make the call for the id manually when the `signTransaction` method gets executed? I'm not sure about that...

Reply:
Wait, I'm confused -- HDWalletProvider wraps another provider, right?  That it forwards requests to, and sometimes translates?  E.g., it translates `eth_sendTransaction` into `eth_sendRawTransaction`?  So it can already interact with that provider, and do an `eth_chainId`?  Why would you need to add a new library?

Reply:
This is a bit confusing to me because the `signMessage` method is defined in the constructor. I think the provider gets added later and `signMessage` won't have access to the context it seems. I'll have to do a little more research and see if there is a way we could possibly rearrange it or if we even might somehow have access.

Reply:
Well, I think `signMessage` is supposed to just be for *signing* the message.  Presumably you'd add `chainId` somewhere outside of it.  Remember, adding the chain ID and signing are two separate things!

Reply:
Yes of course, but the problem is I don't know where we can fetch the chain id because we can't do it in the constructor.

Reply:
How would that happen?  And is that even a real problem?  Like, again, we can think of adding the chain ID as something entirely separate from signing.  Adding the chain ID is something we need to do if we receive an RPC request that requires signing, right, such as `eth_signTransaction` or `eth_sendTransaction`, right?  If HDWalletProvider receives one of those, we can add the chain ID, and then sign.  If someone truly wants to manually invoke signing, rather than sending an `eth_signTransaction` or `eth_sendTransaction`, well, maybe it's OK if the chain ID isn't added?

Reply:
But where will we add the fetching? HDWalletProvider is instantiated and then passed to web3. Then web3 uses it to sign when it needs to. There is no room in between the constructor and when `signMessage` gets called to add the chain id. Hypothetically we could check the provider for some kind of custom method to fetch the chain id in Truffle (like if it has the `init` method or something like that) but that feels pretty janky.

Reply:
OK, I went and looked at the code some more and I think I see now what I was misunderstanding.

I thought the room to add this extra logic was, y'know, in the request-response loop.  What I did not understand was that we do not manually provide that loop, but rather rely on provider-engine for it.  Since we don't actually control the request-response loop per se, it looks to me like you're right; there's nowhere to add it.

So, I guess this is OK.  It's kind of disappointing that we're relying on the user to provide this, but, oh well, so it is.  I don't know whether @gnidan has any opinions on this, but I don't see any way to avoid this for now.

Hopefully eventually OpenEthereum fixes their stuff and we can put this back in `@truffle/contract`, so we don't need to rely on user input for it.

Reply:
Btw, since we're adding a new field to the network config, don't forget to update the docs on trufflesuite.com!